src/aspire/utils/bot_align.py

```diff
@@ -0,0 +1,271 @@
+        If alignment fails try larger values.
+    :param refine: Whether to perform refinement. Boolean, defaults True.
+    :param reflect: Whether to address handedness. Boolean, defaults False.
+        If set true, increase max_iters to 350~400.
+    :param tau: Regularization parameter for surrogate problems. Numeric, defaults 1e-3.
+    :param man_max_iter: Stopping criterion for surrogate problems--maximum iterations. Interger, defaults 500.
+    :param man_min_grad: Stopping criterion for surrogate problems--minimum gradient norm. Numeric, defaults 0.1.
+    :param man_min_sz: Stopping criterion for surrogate problems--minimum step size. Numeric, defaults 0.1.
+    :param verbosity: Whether to print surrogate problem optimization details. Boolean, defaults 0.
+    :param dtype: Numeric dtype to perform computations with.
+        Default `None` infers dtype from `vol_ref`.
+    :return: two rotation matrices R_init (without refinement) and R_est (with refinement).
+        If refine = False, then R_est is set equal to R_init.
+    """
+    # Infer dtype
+    dtype = np.dtype(dtype or vol_ref.dtype)
+
+    # Convert volume data dtype if needed.
+    vol_ref = vol_ref.astype(dtype, copy=False)
+    vol_given = vol_given.astype(dtype, copy=False)
+
+    # Store parameters specific to each loss_type.
+    LOSS_TYPES = {
+        "wemd": dict(lengthscale=0.75, wavelet="sym3", level=6),
+        "eu": dict(lengthscale=1),
+    }
+    loss_type = loss_type.lower()
+    if loss_type not in LOSS_TYPES.keys():
+        raise ValueError(f"BOTalign `loss_type` must be one of {LOSS_TYPES.keys()}")
+
+    # Lookup our params
+    loss_params = LOSS_TYPES[loss_type]
+
+    # Downsample Volumes
+    vol_ref_ds = vol_ref.downsample(downsampling_level)
+    vol_given_ds = vol_given.downsample(downsampling_level)
+
+    if loss_type == "wemd":
+        embed_0 = wemd_embed(
+            vol_ref_ds.asnumpy()[0], loss_params["wavelet"], loss_params["level"]
+        )
+
+    def loss(R):
+        """
+        Alignment loss function.
+        """
+        v = vol_given_ds.rotate(Rotation(R)).asnumpy()[0]
+        if loss_type == "eu":
+            return norm(v - vol_ref_ds)
+        elif loss_type == "wemd":
+            embed_v = wemd_embed(v, loss_params["wavelet"], loss_params["level"])
+            return norm(embed_v - embed_0, ord=1)
+
+    def cf(x1, x2):
+        """
+        The squared exponential covariance function over SO(3).
+        """
+        d = norm(x1 - x2) / loss_params["lengthscale"]
+        return np.exp(-(d**2) / 2, dtype=dtype)
+
+    def cf_grad(x1, x2):
+        """
+        Gradient of the squared exponential covariance function.
+        """
+        return cf(x1, x2) * (x2 - x1) / loss_params["lengthscale"] ** 2
+
+    R = np.zeros((max_iters, 3, 3), dtype=dtype)
+    R[0] = np.eye(3, dtype=dtype)
+    if reflect:
+        manifold = pymanopt.manifolds.Stiefel(3, 3)
```

Comment:
Probably, let me see if I can cook one up quickly by enabling and flipping the volume.

Reply:
Hi @RuiyiYang , what is your expectation testing this code running with `reflect` enabled.  I tried a few things and was able to add a test, but I'm uncertain about the results. Thanks

Reply:
To be honest I didn't run through experiments on this part. But based on my experience, there could be a higher chance of failure if the total number of iterations (`max_iters`) is insufficient. So you will see that the final angle_rec is 0 sometimes and around 180 others. I think the crucial thing is that `max_iters` would have to double, say to 350~400 or even more. The reason is that the search space now becomes O(3) which is twice as large as SO(3). 

With that being said, this it is not much different from the usual way of reflecting the volumes first and run the algorithm again (with `reflect=false`), except being a little more automatic. So I would actually not recommend too much setting `reflect=true` in practice.

Reply:
I see. I think the algorithm may be doing the right thing but I was not computing `angle_rec` correctly.

Currently it is comparing the relative angle between `R_rec` and `R_true.T`, where `R_true.T` is just the true rotation (without composing with the reflection). When `reflect` is activated, the returned `R_rec` also includes the reflection. So the `get_angle` function would indeed give nonsense because we are essentially comparing a rotation matrix R and its reflection -R. So I guess something like `get_angle(R_rec,flip@R_true.T)` would be the correct measure, with the correct positioning of the flip matrix (which is always the most difficult part!). 

Anyway, we might also consider removing the `reflect` option. The only place it shows up is in setting the manifold.

Reply:
Actually, that wasn't the issue, I was already comparing with the reflected R_true :).

I was able to get the test to pass by applying the known flip _and_ a 180* rotation ( the result being a different flip than I was expecting).  Neat. This worked for three different random rotations.

```
>>> np.array([[1, 0, 0], [0, 1, 0], [0, 0, -1]]) @ np.array([[-1, 0, 0], [0, 1, 0], [0, 0, -1]])
array([[-1,  0,  0],
       [ 0,  1,  0],
       [ 0,  0,  1]])
```

I don't really understand why that would be. We can run it by someone else to decide if we should delete it or ship it :shipit: .  Thanks

Reply:
Discussed in our dev meeting and decided to remove it.  We can also add it back if you want to revisit.